src/data/buffer/parser/basic.lean

```diff
@@ -284,202 +290,203 @@ lemma guard_eq_fail {p : Prop} [decidable p] :
-  ∀ {reps : ℕ}, (foldr_core f p b reps).valid
-| 0          := failure
-| (reps + 1) := orelse (hp.bind (λ _, foldr_core.bind (λ _, pure))) pure
+instance eof : eof.mono :=
+mono.decorate_error
 
-lemma foldr {f : α → β → β} (hp : p.valid) : valid (foldr f p b) :=
-λ _ _, foldr_core hp _ _
-
-lemma foldl_core_zero {f : β → α → β} {b : β} : (foldl_core f b p 0).valid :=
-failure
+instance foldr_core {f : α → β → β} {b : β} [p.mono] :
+  ∀ {reps : ℕ}, (foldr_core f p b reps).mono
+| 0          := mono.failure
+| (reps + 1) := begin
+  convert mono.orelse,
+  { convert mono.bind,
+    { apply_instance },
+    { exact λ _, @mono.bind _ _ _ _ foldr_core _ } },
+  { exact mono.pure }
+end
 
-lemma foldl_core {f : α → β → α} {p : parser β} (hp : p.valid) :
-  ∀ {a : α} {reps : ℕ}, (foldl_core f a p reps).valid
-| _ 0          := failure
-| _ (reps + 1) := orelse (hp.bind (λ _, foldl_core)) pure
+instance foldr {f : α → β → β} [p.mono] : mono (foldr f p b) :=
+⟨λ _ _, mono.foldr_core.le _ _⟩
+
+instance foldl_core {f : α → β → α} {p : parser β} [p.mono] :
+  ∀ {a : α} {reps : ℕ}, (foldl_core f a p reps).mono
+| _ 0          := mono.failure
+| _ (reps + 1) := begin
+  convert mono.orelse,
+  { convert mono.bind,
+    { apply_instance },
+    { exact λ _, foldl_core } },
+  { exact mono.pure }
+end
 
-lemma foldl {f : α → β → α} {p : parser β} (hp : p.valid) : valid (foldl f a p) :=
-λ _ _, foldl_core hp _ _
+instance foldl {f : α → β → α} {p : parser β} [p.mono] : mono (foldl f a p) :=
+⟨λ _ _, mono.foldl_core.le _ _⟩
 
-lemma many (hp : p.valid) : p.many.valid :=
-hp.foldr
+instance many [p.mono] : p.many.mono :=
+mono.foldr
 
-lemma many_char {p : parser char} (hp : p.valid) : p.many_char.valid :=
-hp.many.map
+instance many_char {p : parser char} [p.mono] : p.many_char.mono :=
+mono.map
 
-lemma many' (hp : p.valid) : p.many'.valid :=
-hp.many.and_then eps
+instance many' [p.mono] : p.many'.mono :=
+mono.and_then
 
-lemma many1 (hp : p.valid) : p.many1.valid :=
-hp.map.seq hp.many
+instance many1 [p.mono] : p.many1.mono :=
+mono.seq
 
-lemma many_char1 {p : parser char} (hp : p.valid) : p.many_char1.valid :=
-hp.many1.map
+instance many_char1 {p : parser char} [p.mono] : p.many_char1.mono :=
+mono.map
 
-lemma sep_by1 {sep : parser unit} (hp : p.valid) (hs : sep.valid) : valid (sep_by1 sep p) :=
-hp.map.seq (hs.and_then hp).many
+instance sep_by1 [p.mono] [sep.mono] : mono (sep_by1 sep p) :=
+mono.seq
 
-lemma sep_by {sep : parser unit} (hp : p.valid) (hs : sep.valid) : valid (sep_by sep p) :=
-(hp.sep_by1 hs).orelse pure
+instance sep_by [p.mono] [hs : sep.mono] : mono (sep_by sep p) :=
+mono.orelse
 
-lemma fix_core {F : parser α → parser α} (hF : ∀ (p : parser α), p.valid → (F p).valid) :
-  ∀ (max_depth : ℕ), valid (fix_core F max_depth)
-| 0               := failure
-| (max_depth + 1) := hF _ (fix_core _)
+instance fix_core {F : parser α → parser α} [hF : ∀ (p : parser α), p.mono → (F p).mono] :
```

Comment:
Type class inference will never solve `∀ (p : parser α), p.mono → (F p).mono`, this should be a regular argument. I actually thought we had a linter for this.